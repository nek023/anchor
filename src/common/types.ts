export const ItemType = {
  Bookmark: 'bookmark',
  History: 'history',
  Tab: 'tab',
} as const

export interface BookmarkItem {
  type: typeof ItemType.Bookmark
  title?: string
  url?: string
}

export interface HistoryItem {
  type: typeof ItemType.History
  title?: string
  url?: string
}

export interface TabItem {
  type: typeof ItemType.Tab
  favIconUrl?: string
  title?: string
  url?: string
  tabIndex: number
  windowId: number
}

export type Item = BookmarkItem | HistoryItem | TabItem
