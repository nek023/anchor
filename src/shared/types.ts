export const ItemType = {
  Bookmark: "bookmark",
  History: "history",
  Tab: "tab",
} as const;

export interface BookmarkItem {
  id: string;
  type: typeof ItemType.Bookmark;
  title?: string;
  url?: string;
}

export interface HistoryItem {
  id: string;
  type: typeof ItemType.History;
  title?: string;
  url?: string;
}

export interface TabItem {
  id: string;
  type: typeof ItemType.Tab;
  faviconUrl?: string;
  title?: string;
  url?: string;
  tabIndex: number;
  windowId: number;
}

export type Item = BookmarkItem | HistoryItem | TabItem;
