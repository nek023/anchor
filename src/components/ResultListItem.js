import React from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import { Result } from '../models';

export default class ResultListItem extends React.Component {
  static propTypes = {
    onClick: PropTypes.func,
    onMouseEnter: PropTypes.func,
    result: Result.propTypes.isRequired,
    selected: PropTypes.bool.isRequired
  }

  static defaultProps = {
    selected: false
  }

  componentDidUpdate = () => {
    if (this.props.selected) {
      this.node.scrollIntoViewIfNeeded(false);
    }
  }

  handleMouseEnter = () => {
    const { onMouseEnter } = this.props;
    if (onMouseEnter) onMouseEnter(this.props.result.index);
  }

  handleClick = () => {
    const { onClick } = this.props;
    if (onClick) onClick(this.props.result);
  }

  shouldComponentUpdate = (nextProps) => {
    return !(
      Immutable.is(this.props.result, nextProps.result) &&
      this.props.selected === nextProps.selected
    );
  }

  render = () => {
    const { result, selected } = this.props;
    const className = selected ? 'item selected' : 'item';

    return (
      <li className={className}
        onMouseEnter={this.handleMouseEnter}
        onClick={this.handleClick}
        ref={(node) => { this.node = node; }}>
        <div className='item-left'>
          <img className='favicon' src={result.item.getFavIconUrl()} />
        </div>
        <div className='item-body'>
          <div className='item-title'>{result.item.title}</div>
          <div className='item-url'>{result.item.url}</div>
        </div>
      </li>
    );
  }
}
