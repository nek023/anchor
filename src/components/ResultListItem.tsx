import React from 'react'
import scrollIntoView from 'scroll-into-view-if-needed'
import styled, { css } from 'styled-components'
import { Item, ItemType } from '../types'

function getFavIconUrl(item: Item) {
  if (item.type === ItemType.Tab && item.favIconUrl) {
    if (item.favIconUrl.startsWith('chrome://theme/')) {
      return 'chrome://favicon'
    }
    return item.favIconUrl
  }
  return `chrome://favicon/${item.url}`
}

const ResultListItemIconContainer = styled.div`
  width: 30px;
  min-width: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
`

const ResultListItemIconImage = styled.img`
  width: 24px;
  height: 24px;
`

const ResultListItemBody = styled.div`
  margin-left: 8px;
  display: flex;
  flex-direction: column;
  overflow: hidden;
  flex: 1;
`

const ResultListItemTitle = styled.div`
  color: #333333;
  font-size: 15px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  flex: 1;
`

const ResultListItemUrl = styled.div`
  color: #9999aa;
  margin-top: 2px
  color: #9999aa
  font-size: 13px
  font-weight: lighter
  overflow: hidden
  text-overflow: ellipsis
  white-space: nowrap
  flex: 1
`

const ResultListItemContainer = styled.li<{ selected: boolean }>`
  width: 100%
  height: 46px
  padding: 3px
  margin-bottom: 2px
  overflow: hidden
  display: flex
  flex-direction: row

  ${(props) =>
    props.selected && css`
      background-color: #0067ff;

      ${ResultListItemTitle} {
        color: #fafafa;
      }

      ${ResultListItemUrl} {
        color: #bcd5fb;
      }
    `
  }
`

type OptionalProps = {
  onClick?: (index: number) => void
  selected: boolean
}

type Props = OptionalProps & {
  index: number
  item: Item
}

export default class ResultListItem extends React.PureComponent<Props> {
  public static defaultProps: OptionalProps = {
    selected: false,
  }

  private listRef = React.createRef<HTMLLIElement>()

  componentDidUpdate() {
    if (this.props.selected && this.listRef.current) {
      scrollIntoView(this.listRef.current, {
        block: 'nearest',
        inline: 'nearest',
        scrollMode: 'if-needed',
      })
    }
  }

  render() {
    return (
      <ResultListItemContainer onClick={this.handleClick} ref={this.listRef}
        selected={this.props.selected}>
        <ResultListItemIconContainer>
          <ResultListItemIconImage src={getFavIconUrl(this.props.item)} />
        </ResultListItemIconContainer>
        <ResultListItemBody>
          <ResultListItemTitle>
            {this.props.item.title}
          </ResultListItemTitle>
          <ResultListItemUrl>
            {this.props.item.url}
          </ResultListItemUrl>
        </ResultListItemBody>
      </ResultListItemContainer>
    )
  }

  private handleClick = () => {
    if (this.props.onClick) {
      this.props.onClick(this.props.index)
    }
  }
}
