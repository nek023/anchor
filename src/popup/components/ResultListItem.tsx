import React, { useCallback, useEffect, useMemo, useRef } from "react";
import scrollIntoView from "scroll-into-view-if-needed";
import { Item, ItemType } from "../../shared/types";
import { getFaviconUrl } from "../lib/getFaviconUrl";

interface ResultListItemProps {
  item: Item;
  selected: boolean;
  onClick?: (item: Item) => void;
}

export const ResultListItem: React.FC<ResultListItemProps> = ({
  item,
  onClick,
  selected,
}) => {
  const itemContainerRef = useRef<HTMLDivElement>(null);

  const handleClick = useCallback(() => {
    if (onClick == null) return;
    onClick(item);
  }, [item, onClick]);

  useEffect(() => {
    if (selected && itemContainerRef?.current != null) {
      scrollIntoView(itemContainerRef?.current, {
        block: "nearest",
        inline: "nearest",
        scrollMode: "if-needed",
      });
    }
  }, [selected]);

  const faviconUrl = useMemo(() => {
    let faviconUrl: string | undefined = undefined;
    if (item.type === ItemType.Tab) faviconUrl = item.faviconUrl;
    if (faviconUrl == null && item.url != null)
      faviconUrl = getFaviconUrl(item.url);
    return faviconUrl;
  }, [item]);

  return useMemo(
    () => (
      <div
        className={`w-full min-h-[3rem] px-1 pb-px flex flex-row items-center gap-2 ${
          selected ? "bg-blue-700" : ""
        }`}
        onClick={handleClick}
        ref={itemContainerRef}
      >
        <div className="min-w-[2rem] flex justify-center">
          <img className="w-6 h-6" src={faviconUrl} alt={item.title} />
        </div>
        <div className="grow flex flex-col gap-0.5 overflow-hidden">
          <div
            className={`text-sm truncate ${
              selected ? "text-gray-100" : "text-gray-700"
            }`}
          >
            {item.title}
          </div>
          <div
            className={`text-xs font-light truncate ${
              selected ? "text-gray-300" : "text-gray-400"
            }`}
          >
            {item.url}
          </div>
        </div>
      </div>
    ),
    [faviconUrl, handleClick, item, selected],
  );
};
