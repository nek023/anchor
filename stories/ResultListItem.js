import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { ResultListItem } from '../src/components';
import { ItemTypes } from '../src/constants';
import { createItem, createResult } from '../test/helpers';
import favicon from './static/favicon.png';

const item = createItem({
  type: ItemTypes.TAB,
  favIconUrl: favicon
});
const result = createResult({ item });

const actions = {
  onClick: action('onClick'),
  onMouseEnter: action('onMouseEnter')
};

storiesOf('ResultListItem', module)
  .add('default', () => (
    <ResultListItem result={result} {...actions} />
  ))
  .add('selected', () => (
    <ResultListItem result={result} selected={true} {...actions} />
  ));
