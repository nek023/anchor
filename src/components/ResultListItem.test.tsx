import { shallow } from 'enzyme'
import React from 'react'
import { BookmarkItem, ItemType } from '../types'
import ResultListItem from './ResultListItem'

describe('ResultListItem', () => {
  let item: BookmarkItem

  beforeEach(() => {
    item = {
      type: ItemType.Bookmark,
      title: 'example',
      url: 'https://example.com',
    }
  })

  it('matches snapshot', () => {
    const wrapper = shallow(<ResultListItem index={0} item={item} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('calls onClick when mouse has been clicked', () => {
    const onClick = jest.fn()
    const wrapper = shallow(
      <ResultListItem index={0} item={item} onClick={onClick} />
    )

    expect(onClick).not.toHaveBeenCalled()

    wrapper.simulate('click')

    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
