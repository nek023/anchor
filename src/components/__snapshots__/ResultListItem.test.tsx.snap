// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultListItem matches snapshot 1`] = `
<styled.li
  onClick={[Function]}
  selected={false}
>
  <styled.div>
    <styled.img
      src="chrome://favicon/https://example.com"
    />
  </styled.div>
  <styled.div>
    <styled.div>
      example
    </styled.div>
    <styled.div>
      https://example.com
    </styled.div>
  </styled.div>
</styled.li>
`;
